File: backend/src/services/documentService.js
const crypto = require('crypto');
const prisma = require('../config/db');
const { logAudit } = require('./auditService');

const ALLOWED_MIME_TYPES = ['application/pdf', 'image/jpeg', 'image/png', 'image/jpg'];
const MAX_FILE_SIZE_BYTES = 5 * 1024 * 1024; // 5 MB

/**
 * Validates document upload metadata.
 */
function validateDocument({ mimeType, fileSize, fileName }) {
  if (!ALLOWED_MIME_TYPES.includes(mimeType?.toLowerCase())) {
    throw new Error(`Unsupported document type (${mimeType}). Allowed types: PDF, JPG, PNG.`);
  }

  if (fileSize > MAX_FILE_SIZE_BYTES) {
    throw new Error(`File size (${(fileSize / (1024 * 1024)).toFixed(2)} MB) exceeds maximum 5 MB limit.`);
  }

  const ext = fileName?.split('.').pop()?.toLowerCase();
  if (!['pdf', 'jpg', 'jpeg', 'png'].includes(ext)) {
    throw new Error('Invalid file extension.');
  }

  return true;
}

/**
 * Generates an S3 storage path key for a staff document.
 */
function generateS3Key({ category, staffProfileId, fileName }) {
  const sanitizedName = fileName.replace(/[^a-zA-Z0-9._-]/g, '_');
  const cat = (category || 'general').toLowerCase();
  return `staff/${cat}/${staffProfileId}/${Date.now()}_${sanitizedName}`;
}

/**
 * Generates a secure, short-lived signed URL for document retrieval.
 * Uses AWS S3 SDK when deployed in AWS or a time-limited cryptographically HMAC-signed URL locally.
 */
function generateSignedUrl({ fileUrl, documentId, expiresInSeconds = 900 }) {
  // If AWS S3 credentials are configured in environment, could use AWS SDK S3Client.
  // For local development and fallback, generate a 15-minute HMAC token.
  const expiresAt = Math.floor(Date.now() / 1000) + expiresInSeconds;
  const secret = process.env.JWT_SECRET || 'hms_secure_document_secret_2026';
  const hmac = crypto.createHmac('sha256', secret);
  hmac.update(`${documentId}:${fileUrl}:${expiresAt}`);
  const signature = hmac.digest('hex');

  // In production with S3 bucket: `https://${process.env.S3_BUCKET_NAME}.s3.amazonaws.com/${fileUrl}?X-Amz-Signature=...`
  // For API serving with access control:
  return `/api/staff/documents/view/${documentId}?expires=${expiresAt}&sig=${signature}`;
}

/**
 * Verifies a time-limited signed URL token.
 */
function verifySignedUrl({ documentId, fileUrl, expires, signature }) {
  const now = Math.floor(Date.now() / 1000);
  if (parseInt(expires) < now) {
    throw new Error('Document signed URL has expired');
  }

  const secret = process.env.JWT_SECRET || 'hms_secure_document_secret_2026';
  const hmac = crypto.createHmac('sha256', secret);
  hmac.update(`${documentId}:${fileUrl}:${expires}`);
  const expectedSig = hmac.digest('hex');

  if (signature !== expectedSig) {
    throw new Error('Invalid document signature');
  }

  return true;
}

module.exports = {
  validateDocument,
  generateS3Key,
  generateSignedUrl,
  verifySignedUrl,
  ALLOWED_MIME_TYPES,
  MAX_FILE_SIZE_BYTES
};

File: backend/app/services/document_service.py
import hmac
import hashlib
import time
import re
from app.config import settings

ALLOWED_MIME_TYPES = ['application/pdf', 'image/jpeg', 'image/png', 'image/jpg']
MAX_FILE_SIZE_BYTES = 5 * 1024 * 1024 # 5 MB

def validate_document(mime_type: str, file_size: int, file_name: str) -> bool:
    if mime_type.lower() not in ALLOWED_MIME_TYPES:
        raise ValueError(f"Unsupported document type ({mime_type}). Allowed types: PDF, JPG, PNG.")

    if file_size > MAX_FILE_SIZE_BYTES:
        raise ValueError(f"File size ({(file_size / (1024 * 1024)):.2f} MB) exceeds maximum 5 MB limit.")

    ext = file_name.split('.')[-1].lower() if '.' in file_name else ''
    if ext not in ['pdf', 'jpg', 'jpeg', 'png']:
        raise ValueError("Invalid file extension.")

    return True

def generate_s3_key(category: str, staff_profile_id: str, file_name: str) -> str:
    sanitized_name = re.sub(r'[^a-zA-Z0-9._-]', '_', file_name)
    cat = (category or 'general').lower()
    return f"staff/{cat}/{staff_profile_id}/{int(time.time() * 1000)}_{sanitized_name}"

def generate_signed_url(file_url: str, document_id: str, expires_in_seconds: int = 900) -> str:
    expires_at = int(time.time()) + expires_in_seconds
    secret = settings.JWT_SECRET.encode('utf-8')
    msg = f"{document_id}:{file_url}:{expires_at}".encode('utf-8')
    sig = hmac.new(secret, msg, hashlib.sha256).hexdigest()
    return f"/api/staff/documents/view/{document_id}?expires={expires_at}&sig={sig}"

def verify_signed_url(document_id: str, file_url: str, expires: str, signature: str) -> bool:
    now = int(time.time())
    if int(expires) < now:
        raise ValueError("Document signed URL has expired")

    secret = settings.JWT_SECRET.encode('utf-8')
    msg = f"{document_id}:{file_url}:{expires}".encode('utf-8')
    expected_sig = hmac.new(secret, msg, hashlib.sha256).hexdigest()

    if not hmac.compare_digest(signature, expected_sig):
        raise ValueError("Invalid document signature")

    return True
